storage: always handle Raft Ready before quiesce request

nvb opened the pull request:
Fixes #19082.

The test was failing when a leader applied a request and immediately quiesced.
This was because the followers were batching the processing of the commit
`MsgApp` (do we have a better name for this?) and the quiesce message. Since we
now defer the handling of Raft's `Ready` struct to the last request in a batch
(#19056), we were letting the quiesce message decide if both messages should
handle the `Ready` state. Since the quiesce message returned early because it
doesn't need to handle the state, neither request ended up doing so. This
resulted in the entry never applying.

Right now this is a quick way to fix this specific flake, but there are other
places where `processRaftRequest` can return early (with and without an error)
and that may indicate that there are other times where we still need to handle
Raft's `Ready` struct. We might need to maintain the condition
`deferringRaftReadyHandling` state some other way to assure that it's performed
whenever any requests in the batch want it.

nvb committed:
storage: always handle Raft Ready before when batching requestd

Fixes #19082.

The test was failing when a leader applied a request and immediately quiesced.
This was because the followers were batching the processing of the commit
`MsgApp` (do we have a better name for this?) and the quiesce message. Since we
now defer the handling of Raft's `Ready` struct to the last request in a batch
(#19056), we were letting the quiesce message decide if both messages should
handle the `Ready` state. Since the quiesce message returned early because it
doesn't need to handle the state, neither request ended up doing so. This
resulted in the entry never applying.

This commit addresses this by lifting `raft.Ready` handling to
`processRequestQueue`. It does so by first splitting up `processRaftRequest`
and `processRaftSnapshotRequest`. It then pulls out a `withReplicaForRequest`
helper. These two make this a lot more readable because snapshot processing
logic is now isolated from non-snapshot processing logic. Finally, the
change pushes the responsibility of calling `handleRaftReady` to the
caller of `processRaftRequestWithReplica`. The reason for doing this is
that we want `processRequestQueue` to have full control over when the
call to handle `raft.Ready` is made.

There are some complications here with `getOrCreateReplica`, since it is
technically called per `RaftMessageRequest` but we're trying to batch
across `RaftMessageRequests`. The assumption I made here was that
`handleRaftReady` should be called on the last `Replica` successfully
retrieved with `getOrCreateReplica`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19164)
<!-- Reviewable:end -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/store.go, line 3560 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-Kw6zY_3yPSe6vGr9R1Y:-Kw6zY_3yPSe6vGr9R1Z:b-55om3e) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fbc9b7fbb3e2f414abb8d6ca29af8b85c668080/pkg/storage/store.go#L3560)):*
> ```Go
> 		// handle raft ready so that we're up-to-date when we quiesce.
> 		last := i == len(infos)-1
> 		handleRaftReady := last || infos[i+1].req.Quiesce
> ```

I think it would be cleaner if we checked in `processRaftRequest` for `Replica.hasRaftReadyRLocked()` and either called `handleRaftReady` or denied quiescing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Ugh, thanks for tracking this down. You're right that it looks like there are more cases where this could be an issue, so some more discipline wouldn't hurt even if this is the only case that's currently broken. Let me know if you want me to have to clean up my own mess instead of you needing to deal with it.

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-:-Kw71EuQ4dqgFzrfPv4S:b-2vbuwr)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/store.go, line 3560 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-Kw6zY_3yPSe6vGr9R1Y:-Kw9GCER-bj9n-O7Q8g3:byttujf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fbc9b7fbb3e2f414abb8d6ca29af8b85c668080/pkg/storage/store.go#L3560)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think it would be cleaner if we checked in `processRaftRequest` for `Replica.hasRaftReadyRLocked()` and either called `handleRaftReady` or denied quiescing.
</blockquote></details>

I addressed this by lifting the `handleRaftReady` logic to `processRequestQueue`. PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I'm generally happy with this refactor. The closure pattern is something I'm not sure about but it seems to work well enough in these cases and I'm not sure the alternative (return the replica and rely on the caller to unlock it, similar to `getOrCreateReplica`) is any better.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but let others take a look.

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/store.go, line 2964 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-Kw9Z6JYNq0bE1zWJptf:-Kw9Z6JZXru3dQgHbDO2:bc88v3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af2f2ae1d503668dab9244db2d56fd6b609bb4c/pkg/storage/store.go#L2964)):*
> ```Go
> // processRaftRequest processes the (non-snapshot) Raft request on the request's
> // specified replica. It then handles any updated Raft Ready state.
> func (s *Store) processRaftRequest(ctx context.Context, req *RaftMessageRequest) *roachpb.Error {
> ```

`processRaftRequestAndReady`?

---

*[pkg/storage/store.go, line 3018 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-Kw9ZKyIzTryDOXaXaQQ:-Kw9ZKyIzTryDOXaXaQR:bo7dtt2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af2f2ae1d503668dab9244db2d56fd6b609bb4c/pkg/storage/store.go#L3018)):*
> ```Go
> }
> 
> // processRaftRequest processes the incoming snapshot Raft request on the
> ```

`processRaftSnapshotRequest`

---

*[pkg/storage/store.go, line 3025 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-Kw9_qvTS52nx0sXAoJh:-Kw9_qvTS52nx0sXAoJi:b-qi9mr1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af2f2ae1d503668dab9244db2d56fd6b609bb4c/pkg/storage/store.go#L3025)):*
> ```Go
> ) *roachpb.Error {
> 	return s.withReplicaForRequest(ctx, req, func(
> 		ctx context.Context, r *Replica,
> ```

This closure won't allocate, right?

---

*[pkg/storage/store.go, line 3598 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-Kw9aGzz0CR196b4aN2w:-Kw9aGzz0CR196b4aN2x:ba4ymsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af2f2ae1d503668dab9244db2d56fd6b609bb4c/pkg/storage/store.go#L3598)):*
> ```Go
> 		pErr := s.withReplicaForRequest(info.respStream.Context(), info.req,
> 			func(ctx context.Context, r *Replica) *roachpb.Error {
> 				// Save the last Replic we see, since we don't know in advance which
> ```

Replica

---

*[pkg/storage/store.go, line 3602 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-Kw9b39nQE-U_uRQ63Dd:-Kw9ahqXWvi-Y5M6w8Tj:b4qvq8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af2f2ae1d503668dab9244db2d56fd6b609bb4c/pkg/storage/store.go#L3602)):*
> ```Go
> 				// so we can handle the Raft Ready state all at once.
> 				//
> 				// TODO DURING REVIEW: getOrCreateReplica has some weird error cases.
> ```

Yeah, it seems that the last message could always be "from the past" and wouldn't be able to retrieve the current replica any more. As awkward as the current code looks, I'm not sure it can be simplified (though it could be rearranged).

I wonder how expensive it is to re-lock here. Probably negligible.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-:-Kw9bq-DE0ffjuMcvcTy:bmf5kky)*
<!-- Sent from Reviewable.io -->

petermattis commented:
The use of closures provides nice structure, yet we'd need to benchmark that they don't affect performance. That might be reason enough to avoid that pattern in this change.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-:-KwBNO7W-oH4jY5JFvCX:b-3i777w)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
I'm not seeing any movement on `kv`. I'd like to think a closure (that shouldn't even require allocation) per Raft request would be negligible, but my faith has been shaken in recent weeks (#19078).

---

Review status: 2 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-:-KwBWWc5bQypoqrtuVLk:b-o4nlpi)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but what `--read-percent` did you test `kv` with?

---

Review status: 2 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store.go, line 3602 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-Kw9b39nQE-U_uRQ63Dd:-KwBjkLJWwjfl5xq1zwH:b-gmgkff) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af2f2ae1d503668dab9244db2d56fd6b609bb4c/pkg/storage/store.go#L3602)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

That's the unfortunate conclusion I came to as well. I guess we could optimize the common case and call `handleRaftReadyRaftMuLocked` in `withReplicaForRequest` if we enter the closure for the last `info`.  We would then set `lastRepl` to `nil` and avoid calling `handleRaftReady` below. This it's worth the extra complexity?
</blockquote></details>

Can we just make this assignment to `lastRepl` dependent on `r` being non-nil?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-:-KwBjxe9BDtks2Fz7GXE:b8fc3fe)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
I was testing with 20% reads.

---

Review status: 2 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store.go, line 3602 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-Kw9b39nQE-U_uRQ63Dd:-KwBkqrYtjtgxZ0Qie9C:bgnhrej) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af2f2ae1d503668dab9244db2d56fd6b609bb4c/pkg/storage/store.go#L3602)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Can we just make this assignment to `lastRepl` dependent on `r` being non-nil?
</blockquote></details>

I'm not sure that I follow. How would `r` be nil here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-:-KwBkVJedquJaB8kL_QF:bev1rs1)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: 2 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store.go, line 3602 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-Kw9b39nQE-U_uRQ63Dd:-KwCFW_g19Zkod08lMeB:bul5x40) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af2f2ae1d503668dab9244db2d56fd6b609bb4c/pkg/storage/store.go#L3602)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm not sure that I follow. How would `r` be nil here?
</blockquote></details>

Never mind, I misread things.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164)*
<!-- Sent from Reviewable.io -->

nvb committed:
storage: handleRaftReady without releasing lock when possible

This change adjusts `processRequestQueue` to call `handleRaftReadyRaftMuLocked`
without giving up `raftMu` whenever possible. This is only not possible
when the last request in the `infos` fails to retrieve a `Replica`.

nvanbenschoten commented:
Review status: 1 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/storage/store.go, line 3602 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-Kw9b39nQE-U_uRQ63Dd:-KwCL7tMuM6LkpQlmJ_C:bk35aps) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af2f2ae1d503668dab9244db2d56fd6b609bb4c/pkg/storage/store.go#L3602)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Never mind, I misread things.
</blockquote></details>

I added a second commit to avoid unlocking `raftMu` whenever possible.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19164#-:-KwFRQim5zL1vqmeUmol:bnfp4nl)*
<!-- Sent from Reviewable.io -->
